ldap/servers/plugins/memberof/memberof.c

```diff
@@ -958,6 +1025,10 @@ memberof_postop_modrdn(Slapi_PBlock *pb)
                         slapi_log_err(SLAPI_LOG_ERR, MEMBEROF_PLUGIN_SUBSYSTEM,
                                       "memberof_postop_modrdn - Delete dn callback failed for (%s), error (%d)\n",
                                       slapi_entry_get_dn(post_e), ret);
+                    } else {
+                        /* Need to check subtree exclusion and clean up the
+                         * entry's groups if needed */
+                        memberof_cleanup_after_modrdn(&configCopy, pre_e, post_e);
```

Comment:
I am a bit lost here.
My understanding is that here the post_e gets out of the scope, and this function updates all groups with 'member: post_e.dn'. Is that correct ?

In which case this can happen ? A group have a member that did not yet exist ?

Reply:
Ok, yeah I realize this is confusing.  MemebrOf (not RI plugin) does a lot of manipulation before it hits this code.  By the time we get here the "Post entry" has already been stripped of the memberOf attributes.  The groups the Pre-entry belong to have already had the uniquemember DN reanmed to the Post DN:

    uid=mark,ou=included,o=test
    memberOf: cn=group1,ou=groups,o=test

    cn=group,ou=groups,o=test
    uniquemember: uid=mark,ou=included,o=test

Do the modrdn of uid=mark --->  uid=mark,ou=excluded,o=test

Then in MemberOf modrdn postop when we get to my code the entries look like this:

    uid=mark,ou=excluded,o=test

    cn=group,ou=groups,o=test
    uniquemember: uid=mark,ou=excluded,o=test

So to cleanup the groups the user used to belong to we need to look at the user's "pre-entry" so we can read the "memberOf" attributes and find the groups the "pre entry" belonged to.  But now these groups already had the uniquemember renamed to the new post entry DN.  So we use the "post entry" DN to cleanup the groups that were found from the "pre entry".  

I hope that clears this up a little.  The handling of the excluded subtree is a corner case and the include/exclude feature was added after the MO plugin was originally designed and written.  So I'm sure this could be cleaner, but it would required rewriting the function (risky!).  The safest approach is to just to deal with the quirkiness of the MO behavior, and handle it this way IMHO.

Reply:
Thanks Mark for the explanation, so your change is good to me. ACK.
I have a comment, if at this point groups have a reference to post_E, what is the purpose of the function memberof_del_dn_from_groups few lines above ?

Regarding perf pov, include/exclude is a hit, triggering at least 2 times more internal MOD that strictly required.
Assume the entry E, in include scope, belongs to N groups. Then the entry is move in an exclude scope.
MO updates  the N groups to add 'member: E.post', then updates E to remove membership attribute (like 'member') (memberof_del_attr_list) , then updates E to remove 'memberof' (memberof_del_dn_type_callback), then updates the N groups to remove 'member: E.post'. That is 2*(N+1) updates where in theory, only N+1 were required.
We should revisit MO to make it more efficient and compare with what @Firstyear suggested long time ago.

Reply:
> Thanks Mark for the explanation, so your change is good to me. ACK. I have a comment, if at this point groups have a reference to post_E, what is the purpose of the function memberof_del_dn_from_groups few lines above ?

Thanks for making me revisit this function.  So after further investigation I found that the group contains the post & pre DN's as members.  In the code we were only cleaning up the pre-entry DN, and not the post-entry DN.  So I changed the fix to do both.  Please review...

Reply:
Indeed it is much simpler to read. ACK